packages/flutter/lib/src/material/navigation_bar.dart

```diff
@@ -0,0 +1,1169 @@
+/// Destination Widget for displaying Icons + labels in the Material 3
+/// Navigation Bars through [NavigationBar.destinations].
+///
+/// The destination this widget creates will look something like this:
+/// =======
+/// |
+/// |  ☆  <-- [icon] (or [selectedIcon])
+/// | text <-- [label]
+/// |
+/// =======
+class NavigationDestination extends StatelessWidget {
+  /// Creates a navigation bar destination with an icon and a label, to be used
+  /// in the [NavigationBar.destinations].
+  const NavigationDestination({
+    Key? key,
+    required this.icon,
+    this.selectedIcon,
+    required this.label,
+    this.tooltip,
+  }) : super(key: key);
+
+  /// The [Widget] (usually an [Icon]) that's displayed for this
+  /// [NavigationDestination].
+  ///
+  /// The icon will use [NavigationBarThemeData.unselectedIconTheme]. If this is
+  /// null, the default [IconThemeData] would use a size of 24.0 and
+  /// [ColorScheme.onSurface].
+  final Widget icon;
+
+  /// The optional [Widget] (usually an [Icon]) that's displayed when this
+  /// [NavigationDestination] is selected.
+  ///
+  /// If [selectedIcon] is non-null, the destination will fade from
+  /// [icon] to [selectedIcon] when this destination goes from unselected to
+  /// selected.
+  ///
+  /// The icon will use [NavigationBarThemeData.selectedIconTheme]. If this is
+  /// null, the default [IconThemeData] would use a size of 24.0 and
+  /// [ColorScheme.onSurface].
+  final Widget? selectedIcon;
+
+  /// The text label that appears below the icon of this
+  /// [NavigationDestination].
+  ///
+  /// The accompanying [Text] widget will use
+  /// [NavigationBarThemeData.selectedLabelTextStyle] and
+  /// [NavigationBarThemeData.unselectedLabelTextStyle] appropriately. If these
+  /// are null, the default text style would use [TextTheme.overline] with
+  /// [ColorScheme.onSurface].
+  final String label;
+
+  /// The text to display in the tooltip for this [NavigationDestination], when
+  /// the user long presses the destination.
+  ///
+  /// If [tooltip] is an empty string, no tooltip will be used.
+  ///
+  /// Defaults to null, in which case the [label] text will be used.
+  final String? tooltip;
+
+  @override
+  Widget build(BuildContext context) {
+    final ThemeData theme = Theme.of(context);
+    final ColorScheme colorScheme = theme.colorScheme;
+    final NavigationBarThemeData navigationBarTheme = NavigationBarTheme.of(context);
+    final Animation<double> animation = NavigationDestinationInfo.of(context).selectedAnimation;
+
+    return _NavigationDestinationBuilder(
+      label: label,
+      tooltip: tooltip,
+      buildIcon: (BuildContext context) {
+        final IconThemeData defaultIconTheme = IconThemeData(
+          size: 24,
+          color: colorScheme.onSurface,
+        );
+        final Widget selectedIconWidget = IconTheme.merge(
+          data: navigationBarTheme.iconTheme?.resolve(<MaterialState>{MaterialState.selected}) ?? defaultIconTheme,
```

Comment:
It would be nice if one could vary the iconTheme or labelTextStyle based on MaterialState.hovered. It would be OK to leave that out of this PR though.

Reply:
yeah thats a nice-to-have that could be done in a follow-up if requested by users